state/txindex/kv/kv.go

```diff
@@ -46,13 +59,321 @@ func (txi *TxIndex) Get(hash []byte) (*types.TxResult, error) {
+	// no iterate over kvs that are not within range.
+	ranges, rangeIndexes := lookForRanges(conditions)
+	if len(ranges) > 0 {
+		skipIndexes = append(skipIndexes, rangeIndexes...)
+
+		for _, r := range ranges {
+			if !hashesInitialized {
+				hashes = txi.matchRange(r, startKeyForRange(r, height))
+				hashesInitialized = true
+			} else {
+				hashes = intersect(hashes, txi.matchRange(r, startKeyForRange(r, height)))
+			}
+		}
+	}
+
+	// for all other conditions
+	for i, c := range conditions {
+		if intInSlice(i, skipIndexes) {
+			continue
+		}
+
+		if !hashesInitialized {
+			hashes = txi.match(c, startKey(c, height))
+			hashesInitialized = true
+		} else {
+			hashes = intersect(hashes, txi.match(c, startKey(c, height)))
+		}
+	}
+
+	results := make([]*types.TxResult, len(hashes))
+	i := 0
+	for _, h := range hashes {
+		results[i], err = txi.Get(h)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get Tx{%X}", h)
+		}
+		i++
+	}
+
+	return results, nil
+}
+
+func lookForHash(conditions []query.Condition) (hash []byte, err error, ok bool) {
+	for _, c := range conditions {
+		if c.Tag == types.TxHashKey {
+			decoded, err := hex.DecodeString(c.Operand.(string))
+			return decoded, err, true
+		}
+	}
+	return
+}
+
+func lookForHeight(conditions []query.Condition) (height uint64, index int) {
+	for i, c := range conditions {
+		if c.Tag == types.TxHeightKey {
+			return uint64(c.Operand.(int64)), i
+		}
+	}
+	return 0, -1
+}
+
+type queryRanges map[string]queryRange
+
+type queryRange struct {
+	key               string
+	lowerBound        interface{} // int || time.Time
+	includeLowerBound bool
+	upperBound        interface{} // int || time.Time
+	includeUpperBound bool
+}
+
+func lookForRanges(conditions []query.Condition) (ranges queryRanges, indexes []int) {
+	ranges = make(queryRanges)
+	for i, c := range conditions {
+		if isRangeOperation(c.Op) {
+			r, ok := ranges[c.Tag]
+			if !ok {
+				r = queryRange{key: c.Tag}
+			}
+			switch c.Op {
+			case query.OpGreater:
+				r.lowerBound = c.Operand
+			case query.OpGreaterEqual:
+				r.includeLowerBound = true
+				r.lowerBound = c.Operand
+			case query.OpLess:
+				r.upperBound = c.Operand
+			case query.OpLessEqual:
+				r.includeUpperBound = true
+				r.upperBound = c.Operand
+			}
+			ranges[c.Tag] = r
+			indexes = append(indexes, i)
+		}
+	}
+	return ranges, indexes
+}
+
+func isRangeOperation(op query.Operator) bool {
+	switch op {
+	case query.OpGreater, query.OpGreaterEqual, query.OpLess, query.OpLessEqual:
+		return true
+	default:
+		return false
+	}
+}
+
+func (txi *TxIndex) match(c query.Condition, startKey []byte) (hashes [][]byte) {
+	if c.Op == query.OpEqual {
+		it := txi.store.IteratorPrefix(startKey)
+		defer it.Release()
+		for it.Next() {
+			hashes = append(hashes, it.Value())
+		}
+	} else if c.Op == query.OpContains {
+		// XXX: doing full scan because startKey does not apply here
```

Comment:
@ebuchman do you still don't understand why we need to do full scan here?

```
like startKey = "account.owner=an"
search query = "accoutn.owner CONSISTS an"
```

we can't iterate with prefix `account.owner=an` because we might miss keys like `account.owner=Ulan`